sql/analyzer/scope.go

```diff
@@ -25,18 +25,48 @@ type Scope struct {
 	// Stack of nested node scopes, with innermost scope first. A scope node is the node in which the subquery is
 	// defined, or an appropriate sibling.
 	nodes []sql.Node
+	// Memo nodes are nodes in the execution context that shouldn't be considered for name resolution, but are still
+	// important for analysis.
+	memos []sql.Node
 }
 
 // newScope creates a new Scope object with the additional innermost Node context. When constructing with a subquery,
 // the Node given should be the sibling Node of the subquery.
 func (s *Scope) newScope(node sql.Node) *Scope {
 	if s == nil {
-		return &Scope{[]sql.Node{node}}
+		return &Scope{nodes: []sql.Node{node}}
 	}
 	var newNodes []sql.Node
 	newNodes = append(newNodes, node)
 	newNodes = append(newNodes, s.nodes...)
-	return &Scope{newNodes}
+	return &Scope{nodes: newNodes, memos: s.memos}
+}
+
+// memo creates a new Scope object with the memo node given. Memo nodes don't affect name resolution, but are used in
+// other parts of analysis, such as error handling for trigger / procedure execution.
+func (s *Scope) memo(node sql.Node) *Scope {
+	if s == nil {
+		return &Scope{memos: []sql.Node{node}}
+	}
+	var newNodes []sql.Node
```

Comment:
From looking at the rest of the code, it looks like we want to prepend any given node as we want the innermost node first when looping over the slice, and we assume that we're giving nodes outer to inner here. Perhaps memos are only intended for name resolution so it doesn't matter, but if not (and assuming I'm not misreading anything) we should add a comment saying why we're prepending here.

Reply:
Doesn't really matter, just making it consistent with the storage order for scope nodes (inner to outer)